webrender/src/device/gfx/device.rs

```diff
@@ -1282,14 +1290,50 @@ impl<B: hal::Backend> Device<B> {
                 &program
                     .pipelines
                     .get(&(format, blend_state, self.render_pass_depth_state, self.current_depth_test))
-                    .expect(&format!(
-                        "The blend state {:?} with depth test {:?} and render_pass_depth_state {:?} and format {:?} not found for {} program!",
-                        blend_state, self.current_depth_test, self.render_pass_depth_state, format, program.shader_name
-                    )),
+                    .expect(
+                        // Creating this string takes a lot of time, so we only enable it in debug
+                        #[cfg(debug_assertions)]
+                        &format!(
+                            "The blend state {:?} with depth test {:?} and render_pass_depth_state {:?} and format {:?} not found for {} program!",
+                            blend_state, self.current_depth_test, self.render_pass_depth_state, format, program.shader_name
+                        ),
+                        #[cfg(not(debug_assertions))]
+                        "Pipeline not found"
+                    ),
             );
             if let Some(SUBPIXEL_CONSTANT_TEXT_COLOR) = blend_state  {
                 self.command_buffer.set_blend_constants(self.blend_color.to_array());
             }
+            if self.pipeline_layout_changed || self.rebind_descriptors {
```

Comment:
it feels to me that there is some redundancy between `pipeline_layout_changed` and `rebind_descriptors`. Can there be only one?

Reply:
Yes we can!

Reply:
Removed it in https://github.com/szeged/webrender/pull/350/commits/b0be810ff981417d41a2600b6a4cd74ae5f9a2af#diff-cdf03c96bd34cb98c2d09b0be7658232L355